### ISSUE TITLE
[css-fonts-4] Consider always serializing the angle for `font-style: oblique`

### ISSUE THREAD

**Issue by nt1m** [w3c/csswg-drafts#8291]
Right now, the angle should only be serialized "if specified" according to the spec. 
http://w3c.github.io/csswg-drafts/css-fonts-4/#font-style-prop

Given the default angle has changed from 20deg to 14deg, I think it is better to always serialize the angle for backwards compatibility, since a website may expect oblique to mean either.

**Comment by nt1m** [comment]
cc @litherum @jfkthame @svgeesus @lilles @emilio What do you think?

**Comment by litherum** [comment]
I think it's a great idea! I don't see a downside (other than it kind of violates the shortest serialization principle), but being explicit about what oblique means seems like a good idea. Especially if it changes in the future!

**Comment by jfkthame** [comment]
It would mean that the serialized form would be incompatible with older UAs that only support the `oblique` keyword, but not the specification of an angle. IMO we should stay with the backward-compatible shortest serialization, and omit the angle when it is the default value.

**Comment by nt1m** [comment]
@jfkthame The problem is that "default angle" means 20deg or 14deg depending on the browser version, and that is problematic for backwards compatibility.

If a website developed for an older UA reads `oblique` they are more likely going to assume `20deg` than `14deg`.

**Comment by svgeesus** [comment]
> Checking the history of CSS Fonts 4, it seems 20deg was given as the default in the [FPWD](https://www.w3.org/TR/2017/WD-css-fonts-4-20170711/#font-style-prop) in 2017. However, it was changed to 14deg already the next year.

Right, although I recall seeing some 20deg assumptions in WPT fairly recently, like in the last year.

I understand the concern about backwards compat with in-the-wild code that was correct with CSS2 or CSS Fonts 3, but also understand the concern about 14deg vs. 20deg which affects more modern code. Not seeing an obvious path forward here.

**Comment by litherum** [comment]
(Up until this week, WebKit used 20deg instead of 14deg.)

**Comment by felipeerias** [comment]
From my point of view, serialising `oblique 14deg` as `oblique` may lose authorial intent.

An author who writes `oblique` is requesting the default (whatever it might be), whereas one who writes `oblique 14deg` (or a `calc()` expression that resolves to it) is requesting a specific angle.

Furthermore, each engine quantises angles to a different internal fixed-point precision and therefore there are unexpected interoperability effects at the margins.

Consider, for example, **`oblique 14.2deg`**. Naively, one might expect that values to serialise to itself, however:

- Chrome computes is as `italic` (non spec-compliant).
- WebKit rounds the angle to `14deg` and computes that value to `oblique`.
- Firefox uses a different rounding and computes the value to `oblique 14.1992deg`.

https://github.com/w3c/csswg-drafts/issues/11633

**Comment by svgeesus** [comment]
@jfkthame @nt1m @felipeerias circling back to this old issue, do we have any web-compat data on sites whose scripts are expecting `oblique` and would break on getting `oblique 14deg` or indeed any explicit angle?

It seems we are held up on uncertainty, here.

**Comment by svgeesus** [comment]
related

 - https://github.com/w3c/csswg-drafts/issues/11633

**Comment by felipeerias** [comment]
@svgeesus @jfkthame @nt1m

Currently [CSS Fonts 4 - 2.4](https://www.w3.org/TR/css-fonts-4/#font-style-prop) states:

> Name: 	font-style
> …
> Computed value: 	the keyword specified, plus angle in degrees if specified

My reading of that is that the computed value should preserve the author's intent:

- `font-style: oblique` computes as `oblique`
- `font-style: oblique 14deg` computes as `oblique 14deg`

If it is decided to follow this route, perhaps the spec should specify that the lack of an `<angle>` represents `14deg` at used-value time.

**Comment by svgeesus** [comment]
> My reading of that is that the computed value should preserve the author's intent:
> 
>     * `font-style: oblique` computes as `oblique`
> 
>     * `font-style: oblique 14deg` computes as `oblique 14deg`
> 
> 
> If it is decided to follow this route, perhaps the spec should specify that the lack of an `<angle>` represents `14deg` at used-value time.

@astearns that seems both reasonable and web-compatible, perhaps we can get an async resolution to close off this issue?

**Comment by astearns** [comment]
The first bit (what gets computed) seems to me like a straightforward reading of the spec. Should we just have a resolution to add the examples, or is there more needed?

**Comment by felipeerias** [comment]
@astearns @svgeesus Some WPT tests would need to be updated because they currently expect `oblique 14deg` to serialise as `oblique`, for example:

- [/css/css-fonts/parsing/font-style-computed.html](https://github.com/web-platform-tests/wpt/blob/master/css/css-fonts/parsing/font-style-computed.html)
- [/css/css-fonts/variations/font-style-parsing.html](https://github.com/web-platform-tests/wpt/blob/master/css/css-fonts/variations/font-style-parsing.html)

I can take care of that after there is a resolution on this matter.

**Comment by astearns** [comment]
The CSSWG will automatically accept this resolution one week from now if no objections are raised here. Anyone can add an emoji to this comment to express support. If you do not support this resolution, please add a new comment.

Proposed Resolution: Make explicit that `oblique 14deg` computes and serializes to `oblique 14deg`

**Comment by jfkthame** [comment]
The proposed resolution seems fine, as far as it goes; but I wonder if we should also clarify what [the spec](https://drafts.csswg.org/css-fonts/#valdef-font-style-oblique-angle--90deg-90deg) means when it says "The lack of an [`<angle>`](https://drafts.csswg.org/css-values-4/#angle-value) represents 14deg."

Currently, Gecko handles this "lack of an angle represent[ing] 14deg" at [parse time](https://searchfox.org/firefox-main/rev/22adc0417a3dbf478c7af00d45daa311e2efdf8b/servo/components/style/values/specified/font.rs#304-309), so that `oblique` and `oblique 14deg` are internally the exact same value. The intended outcome of the discussion here, AIUI, is effectively saying that "represents 14deg" applies only at _used value_ time; _specified_ and _computed_ values both maintain the distinction between the absence of an angle and the presence of `14deg`.

So I think it would be helpful to rephrase the "represents 14deg" sentence to make it more explicit that this is referring _only_ to the used value. (I'm aware we'll need to update Gecko's style code to handle this; that's fine, I'm not opposed to making such a change.)

**Comment by svgeesus** [comment]
I agree that _represents_ is imprecise and that _has a used value of_ or something like that would be better.

**Comment by astearns** [comment]
RESOLVED: Make explicit that `oblique 14deg` computes and serializes to `oblique 14deg`
